chart/Graphics/Rendering/Chart/Axis/Floating.hs

```diff
@@ -77,14 +78,111 @@ instance PlotValue LogValue where
     fromValue d          = LogValue (exp d)
     autoAxis             = autoScaledLogAxis def
 
+-- | Show a list of axis labels.
+-- If some are too big or all are too small, switch to scientific notation for all.
+-- If the range is much smaller than the mean, use an offset.
+-- TODO: show this offset only once, not on every label.
+-- When thinking about improving this function,
+-- https://github.com/matplotlib/matplotlib/blob/master/lib/matplotlib/ticker.py
+-- is a good read.
+--
+-- >>> showDs [0, 1, 2 :: Double]
+-- ["0","1","2"]
+--
+-- >>> showDs [0, 1000, 2000 :: Double]
+-- ["0.0e0","1.0e3","2.0e3"]
+--
+-- >>> showDs [0, 0.001, 0.002 :: Double]
+-- ["0","0.001","0.002"]
+--
+-- >>> showDs [-10000, -1000, 9000 :: Double]
+-- ["-1.0e4","-1.0e3","9.0e3"]
+--
+-- >>> showDs [10, 11, 12 :: Double]
+-- ["10","11","12"]
+--
+-- >>> showDs [100, 101, 102 :: Double]
+-- ["100","101","102"]
+--
+-- >>> showDs [1000, 1001, 1002 :: Double]
+-- ["1000","1001","1002"]
+--
+-- >>> showDs [10000, 10001, 10002 :: Double]
+-- ["1.0e4 + 0","1.0e4 + 1","1.0e4 + 2"]
+--
+-- >>> showDs [-10000, -10001, -10002 :: Double]
+-- ["-1.0e4 + 2","-1.0e4 + 1","-1.0e4 + 0"]
+showDs :: forall d . (RealFloat d) => [d] -> [String]
+showDs xs
+  | useOffset = map addShownOffset $ showWithoutOffset (map (\x -> x - offset) xs)
+  | otherwise = showWithoutOffset xs
+  where
+    -- if the range is much smaller than the mean
+    -- AND the data is either all positive or all negative, apply an offset
+    useOffset
+      -- if some data is positive and some negative, we don't need an offset
+      | min' <= 0 && max' >= 0 = False
+      -- if the range is significantly smaller than the average, we need an offset
+      | 1000 * abs (max' - min') < abs mean' = True
+      | otherwise = False
+
+    mean' :: d
+    mean'
+      | n == 0 = 0
+      | otherwise = sum xs / fromIntegral n
+      where
+        n = length xs
+    min' = minimum xs
+    max' = maximum xs
+
+    -- Use the min for offset. Another good choice could be the mean.
+    offset :: d
+    offset = min'
+    shownOffset = case showWithoutOffset [offset] of
+      [r] -> r
+      rs -> error $ "showDs: shownOffset expected 1 element, got " ++ show (length rs)
+
+    addShownOffset :: String -> String
+    addShownOffset ('-':x) = shownOffset ++ " - " ++ x
+    addShownOffset x = shownOffset ++ " + " ++ x
+
+showWithoutOffset :: RealFloat d => [d] -> [String]
+showWithoutOffset xs
+  | useScientificNotation = map (\x -> showEFloat' (Just 1) x) xs
+  | otherwise = map showD xs
+  where
+    -- use scientific notation if max value is too big or too small
+    useScientificNotation = maxAbs > 1100 || maxAbs < 0.001
```

Comment:
Why these particular constants? I'd have probably made them more extreme, so the scientific mode isn't shown until e6. Was this your preference, or something more?